### PR TITLE
Fix Indices Request Cache ITs from flaking and timing out

### PR DESCRIPTION
**Pull request by kiranprakash154**
### Description

There were [multiple issues](https://github.com/opensearch-project/OpenSearch/issues?q=is%3Aissue+is%3Aopen+sort%3Aupdated-desc+assignee%3Akiranprakash154+%5BBUG%5D) created for Indices Request Cache's IT tests flaking & the suite timing out. 

## Addressing suite timing out

I believe the reason the tests were timing out because the new stalekey mgmt related tests added are spinning up an internal test cluster for every run (since we want different refresh intervals in some of the tests) and each test is run 4 times due to concurrent segment search and tiered caching parameters [here](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/internalClusterTest/java/org/opensearch/indices/IndicesRequestCacheIT.java#L101-L108)

So i have moved the stalekey mgmt related tests out into its own class since stalekey mgmt is agnostic of cache type underneath and concurrent segment search. 

## Addressing Flakiness
Upon looking into the failures, the tests were failing on cache size assertions. 

I think its due to the background refreshes and segment merges happening that is not in our control. 
Hence I have gone ahead and addressed them in this PR by 
- Disabling Refresh Intervals 
- Disable segment merges

```java
    // Disable index refreshing to avoid cache being invalidated mid-test
    .put(IndexSettings.INDEX_REFRESH_INTERVAL_SETTING.getKey(), TimeValue.timeValueMillis(-1))
	// Disable background segment merges invalidating the cache
    .put(MergePolicyProvider.INDEX_MERGE_ENABLED, false)
```

and also asserting the segment sizes

```java
	// assert segment counts stay the same
	assertEquals(1, getSegmentCount(client, index));
```

### Local Testing

I configured the test suite to run each test 100 times. 

SCR-20240619-lqgz

### Related Issues
Resolves 
https://github.com/opensearch-project/OpenSearch/issues/13711
https://github.com/opensearch-project/OpenSearch/issues/13437
https://github.com/opensearch-project/OpenSearch/issues/13540
https://github.com/opensearch-project/OpenSearch/issues/13503
https://github.com/opensearch-project/OpenSearch/issues/13949
https://github.com/opensearch-project/OpenSearch/issues/14173
https://github.com/opensearch-project/OpenSearch/issues/14242
https://github.com/opensearch-project/OpenSearch/issues/11374
https://github.com/opensearch-project/OpenSearch/issues/14243
https://github.com/opensearch-project/OpenSearch/issues/14288

### Check List
- [x] Functionality includes testing.
- [x] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md), if applicable.
- [x] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose), if applicable.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by kiranprakash154** [comment]
> ❌ Gradle check result for [e76c9b6](https://github.com/opensearch-project/OpenSearch/commit/e76c9b64fdfd7c9ab02910fb2283bef0d8c1f3be): [FAILURE](https://build.ci.opensearch.org/job/gradle-check/41052/)
> 
> Please examine the workflow log, locate, and copy-paste the failure(s) below, then iterate to green. Is the failure [a flaky test](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md?rgh-link-date=2024-06-14T23%3A17%3A58Z#flaky-tests) unrelated to your change?

https://github.com/opensearch-project/OpenSearch/issues/12639

**Comment by dblock** [comment]
@kiranprakash154 Want to describe what this does and mark it ready for CR?

File: server/src/internalClusterTest/java/org/opensearch/indices/IndicesRequestCacheIT.java [hunk 1]
```diff
@@ -644,11 +647,14 @@ public void testProfileDisableCache() throws Exception {
                         .put(IndicesRequestCache.INDEX_CACHE_REQUEST_ENABLED_SETTING.getKey(), true)
                         .put(IndexMetadata.SETTING_NUMBER_OF_SHARDS, 1)
                         .put(IndexMetadata.SETTING_NUMBER_OF_REPLICAS, 0)
+                        // Disable index refreshing to avoid cache being invalidated mid-test
+                        .put(IndexSettings.INDEX_REFRESH_INTERVAL_SETTING.getKey(), TimeValue.timeValueMillis(-1))
                 )
                 .get()
         );
         indexRandom(true, client.prepareIndex(index).setSource("k", "hello"));
-        ensureSearchable(index);
```
Comment (sgup432) [inline server/src/internalClusterTest/java/org/opensearch/indices/IndicesRequestCacheIT.java:651]: Same as above. Lets not remove this.

File: server/src/internalClusterTest/java/org/opensearch/indices/IndicesRequestCacheIT.java [hunk 2]
```diff
@@ -687,11 +693,14 @@ public void testCacheWithInvalidation() throws Exception {
                         .put(IndexMetadata.SETTING_NUMBER_OF_SHARDS, 1)
                         .put(IndexMetadata.SETTING_NUMBER_OF_REPLICAS, 0)
                         .put("index.refresh_interval", -1)
+                        // Disable index refreshing to avoid cache being invalidated mid-test
+                        .put(IndexSettings.INDEX_REFRESH_INTERVAL_SETTING.getKey(), TimeValue.timeValueMillis(-1))
                 )
                 .get()
         );
         indexRandom(true, client.prepareIndex(index).setSource("k", "hello"));
-        ensureSearchable(index);
```
Comment (sgup432) [inline server/src/internalClusterTest/java/org/opensearch/indices/IndicesRequestCacheIT.java:694]: Same as above.

File: server/src/internalClusterTest/java/org/opensearch/indices/IndicesRequestCacheIT.java [hunk 3]
```diff
@@ -1329,11 +1345,26 @@ private void setupIndex(Client client, String index) throws Exception {
...
                         .put(IndexMetadata.SETTING_NUMBER_OF_SHARDS, 1)
                         .put(IndexMetadata.SETTING_NUMBER_OF_REPLICAS, 0)
+                        // Disable index refreshing to avoid cache being invalidated mid-test
+                        .put(IndexSettings.INDEX_REFRESH_INTERVAL_SETTING.getKey(), TimeValue.timeValueMillis(-1))
+                        // settings to reduce the chances of background segment merges invalidating the cache
+                        .put(TieredMergePolicyProvider.INDEX_MERGE_POLICY_SEGMENTS_PER_TIER_SETTING.getKey(), 1000d)
+                        .put(TieredMergePolicyProvider.INDEX_MERGE_POLICY_MAX_MERGE_AT_ONCE_SETTING.getKey(), 2)
+                        .put(TieredMergePolicyProvider.INDEX_MERGE_POLICY_MAX_MERGED_SEGMENT_SETTING.getKey(), "100gb")
+                        .put(MergeSchedulerConfig.MAX_THREAD_COUNT_SETTING.getKey(), 1)
+                        .put(MergeSchedulerConfig.MAX_MERGE_COUNT_SETTING.getKey(), 1)
```
Comment (sgup432) [inline server/src/internalClusterTest/java/org/opensearch/indices/IndicesRequestCacheIT.java:1355]: I don't think we need to use all these settings. Can't we just ensure that there is only a single segment created after force merge or for an index?

File: server/src/internalClusterTest/java/org/opensearch/indices/IndicesRequestCacheIT.java [hunk 4]
```diff
@@ -134,7 +139,8 @@ public void testCacheAggs() throws Exception {
             client.prepareIndex(index).setSource("f", "2014-03-10T00:00:00.000Z"),
             client.prepareIndex(index).setSource("f", "2014-05-13T00:00:00.000Z")
         );
-        ensureSearchable(index);
```
[the remark this answers is not in the document]
Comment (kiranprakash154) [reply server/src/internalClusterTest/java/org/opensearch/indices/IndicesRequestCacheIT.java:137]: @sgup432 ensureSearchable(index) is part of forceMerge 
```java
private void forceMerge(Client client, String index) {
        ForceMergeResponse forceMergeResponse = client.admin().indices().prepareForceMerge(index).setFlush(true).get();
        OpenSearchAssertions.assertAllSuccessful(forceMergeResponse);
        refreshAndWaitForReplication();
        ensureSearchable(index);
    }
```

File: server/src/internalClusterTest/java/org/opensearch/indices/IndicesRequestCacheIT.java (hunk 1 above)
Comment (kiranprakash154) [reply server/src/internalClusterTest/java/org/opensearch/indices/IndicesRequestCacheIT.java:651]: discussed the order of ensureSearchable is important in slack, addressed that in the commit. - [9e3c4c1](https://github.com/opensearch-project/OpenSearch/pull/14369/commits/9e3c4c16a2424142011a1ecae44ed92c752c5c17)

**Comment by kiranprakash154** [comment]
> ❕ Gradle check result for [a4a3c34](https://github.com/opensearch-project/OpenSearch/commit/a4a3c346a301190767d9b6c455229ad10f10bc5c): [UNSTABLE](https://build.ci.opensearch.org/job/gradle-check/41288/)
> 
> Please review all [flaky tests](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md?rgh-link-date=2024-06-19T02%3A05%3A49Z#flaky-tests) that succeeded after retry and create an issue if one does not already exist to track the flaky failure.

AzureBlobStoreRepositoryTests.testSnapshotAndRestore - https://github.com/opensearch-project/OpenSearch/issues/14291

File: server/src/internalClusterTest/java/org/opensearch/indices/IndicesRequestCacheCleanupIT.java [hunk 5]
```diff
@@ -0,0 +1,712 @@
...
+        // assert cache is empty for index2
+        assertEquals(0, getRequestCacheStats(client, index2).getMemorySizeInBytes());
+        // create first cache entry in index2
+        createCacheEntry(client, index2, "hello");
+        assertCacheState(client, index2, 0, 1);
+        assertTrue(getRequestCacheStats(client, index2).getMemorySizeInBytes() > 0);
+
+        // invalidate the cache for index1
+        indexRandom(false, client.prepareIndex(index1).setId("1").setSource("d", "hello"));
+        forceMerge(client, index1);
```
Comment (sgup432) [inline server/src/internalClusterTest/java/org/opensearch/indices/IndicesRequestCacheCleanupIT.java:633]: Can we also check whether entries were cleared from cache after this step?

File: server/src/internalClusterTest/java/org/opensearch/indices/IndicesRequestCacheCleanupIT.java [hunk 6]
```diff
@@ -0,0 +1,712 @@
...
+        // invalidate the cache for index2
+        indexRandom(false, client.prepareIndex(index2).setId("1").setSource("d", "hello"));
+        forceMerge(client, index2);
```
Comment (sgup432) [inline server/src/internalClusterTest/java/org/opensearch/indices/IndicesRequestCacheCleanupIT.java:636]: Same as above. We can verify entries were cleared up from cache after force merge.

File: server/src/internalClusterTest/java/org/opensearch/indices/IndicesRequestCacheCleanupIT.java [hunk 7]
```diff
@@ -0,0 +1,712 @@
...
+        // sleep until cache cleaner would have cleaned up the stale key from index2
+        assertBusy(() -> {
+            // cache cleaner should have cleaned up the stale key from index2 and hence cache should be empty
+            assertEquals(0, getRequestCacheStats(client, index2).getMemorySizeInBytes());
+            // cache cleaner should have only cleaned up the stale entities for index1
+            long currentMemorySizeInBytesForIndex1 = getRequestCacheStats(client, index1).getMemorySizeInBytes();
+            // assert the memory size of index1 to only contain 1 entry added after flushAndRefresh
+            assertEquals(memorySizeForIndex1With1Entries, currentMemorySizeInBytesForIndex1);
+            // cache for index1 should not be empty since there was an item cached after flushAndRefresh
+            assertTrue(currentMemorySizeInBytesForIndex1 > 0);
```
Comment (sgup432) [inline server/src/internalClusterTest/java/org/opensearch/indices/IndicesRequestCacheCleanupIT.java:648]: This is not needed considering we are already doing equality check above. And we already know memorySizeForIndex1With1Entries > 0

File: server/src/internalClusterTest/java/org/opensearch/indices/IndicesRequestCacheCleanupIT.java (hunk 5 above)
Comment (kiranprakash154) [reply server/src/internalClusterTest/java/org/opensearch/indices/IndicesRequestCacheCleanupIT.java:633]: @sgup432 that's what i'm doing in the `assertBusy` step below. 
like this for index1
```java
			// cache cleaner should have only cleaned up the stale entities for index1
            long currentMemorySizeInBytesForIndex1 = getRequestCacheStats(client, index1).getMemorySizeInBytes();
```
and for index2
```java
			// cache cleaner should have cleaned up the stale key from index2 and hence cache should be empty
            assertEquals(0, getRequestCacheStats(client, index2).getMemorySizeInBytes());
```

File: server/src/internalClusterTest/java/org/opensearch/indices/IndicesRequestCacheCleanupIT.java (hunk 7 above)
Comment (kiranprakash154) [reply server/src/internalClusterTest/java/org/opensearch/indices/IndicesRequestCacheCleanupIT.java:648]: makes sense

File: server/src/internalClusterTest/java/org/opensearch/indices/IndicesRequestCacheCleanupIT.java (hunk 6 above)
Comment (kiranprakash154) [reply server/src/internalClusterTest/java/org/opensearch/indices/IndicesRequestCacheCleanupIT.java:636]: same as https://github.com/opensearch-project/OpenSearch/pull/14369#discussion_r1646615868

Comment (sgup432) [reply server/src/internalClusterTest/java/org/opensearch/indices/IndicesRequestCacheCleanupIT.java:636]: Yeah we can ignore this one. Not needed.

File: server/src/internalClusterTest/java/org/opensearch/indices/IndicesRequestCacheCleanupIT.java (hunk 5 above)
Comment (kiranprakash154) [reply server/src/internalClusterTest/java/org/opensearch/indices/IndicesRequestCacheCleanupIT.java:633]: I see what u mean, i like it.

**Review by sgup432 — APPROVED** [review]

**Comment by kiranprakash154** [comment]
@andrross / @sohami this ready for your review.

**Review by andrross — APPROVED** [review]

**Comment by andrross** [comment]
@kiranprakash154 Can you take care of the manual backport?

**Comment by kiranprakash154** [comment]
> @kiranprakash154 Can you take care of the manual backport?

I will, thanks !

**Comment by kiranprakash154** [comment]
Created a manual backport - https://github.com/opensearch-project/OpenSearch/pull/14482

**Pull request merged**